stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -394,6 +580,161 @@ public final class _ExecutionContext {
+  }
+  _RuntimeConfig.traceState = .notTracing
+  traceContext.finalize(traceeBasicName: fnName, outputs: outputs)
+  return traceContext
+}
+
+private extension TensorArrayProtocol {
+  // The returned handles are owned by the caller.
+  var cTensorHandles: [CTensorHandle] {
+    debugLog("Getting \(self._tensorHandleCount) C handles.")
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(
+      capacity: Int(self._tensorHandleCount))
+    debugLog("Unpacking handles into buffer.")
+    _unpackTensorHandles(into: buffer)
+    let status = TF_NewStatus()
+    debugLog("Copying buffer content to output handles.")
+    var output: [CTensorHandle] = []
+    for i in 0..<Int(_tensorHandleCount) {
+      let address = buffer.advanced(by: i)
+      let isConcrete = TFE_TensorHandleIsConcrete(address.pointee) != 0
+      debugLog("""
+                 Copying the \(i)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete).
+                 """)
+      let newHandle = TFE_TensorHandleCopySharingTensor(address.pointee,
+                                                        status)
+      checkOk(status)
+      debugLog("""
+                 Copying the \(i)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete) to C handle create \(newHandle!).
+                 """)
+      output.append(newHandle!)
+    }
+    TF_DeleteStatus(status)
+    return output
+  }
+}
+
+private extension TensorGroup {
+  // The tensors in `input` are NOT owned by this instance.
+  init(_copying input: [CTensorHandle]) {
+    assert(Self._tensorHandleCount == input.count)
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(
+      capacity: input.count)
+    let status = TF_NewStatus()
+    // copy input to buffer
+    for (i, inputTensorHandle) in input.enumerated() {
+      let address = buffer.advanced(by: i)
+      // Each tensor can be symbolic (e.g. when using this API to create a
+      // symbolic input instance to tracee) or concrete (e.g. when creating the
+      // final output of the tracee).
+      let newHandle = TFE_TensorHandleCopySharingTensor(inputTensorHandle,
+                                                        status)
+      checkOk(status)
+      address.initialize(to: newHandle!)
+    }
+    TF_DeleteStatus(status)
+    self.init(_owning: buffer)
+  }
+}
+
+// TODO: Fold this protocol into TensorArrayProtocol.
+// This requires that we move concrete implementation such as
+// Tensor._makeInstance() to TensorGroup.swift.
+public protocol TensorArrayProtocolEnhanced : TensorArrayProtocol {
+  // Create an instance based on `inputs`, which can be symbolic (e.g. when
+  // creating a symbolic input to tracee) or concrete (e.g. when creating a
+  // final output of executing the tracee).
+  func _makeInstance<C: Collection>(owning inputs: C) -> Self
+    where C.Element == CTensorHandle
+}
+
+public func trace<State : TensorArrayProtocolEnhanced,
```

Comment:
@jekbradbury and I discussed earlier and we'd like to name this API `graph(with:in:)`, instead of baking the technically inaccurate notion "trace" into the user's mental model.

If you think this feature is not ready yet, it's also okay to add an underscore to it to make it not show up in code completion as a supported feature.

Reply:
Done.